### ISSUE TITLE
Saving all trained params in a single file

### ISSUE THREAD

**Issue by sidgoyal78** [PaddlePaddle/Paddle#7722]
### Merging all params in a single file

For inference, we will to have 2 files, one for the `programDesc` and one that has all the params together. We look at 1 approach to do this.

#### Understanding save/load ops (C++ side)
- From the [model_format design doc](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/model_format.md), we see some details in the table but it is not super clear. So we will look at the implementation.

To understand the current serialization: we look at `save_op`

- In `save_op` the main work is performed by `SerializeToStream( <ofstream>, <framework::LoDTensor>, .. )` [Code](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/lod_tensor.cc#L235). This function saves a version number, size of LoD and actual LoD data. 

- Then it calls, `SerializeToStream(<ofstream>, <Tensor> ..)` [Code](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/tensor_util.h#L218). This function saves a version number, tensor description as a serialized protobuf, and the actual data.  

The corresponding `load_op` basically does the deserialization accordingly (respecting the ordering in the `save_op`).

#### Understanding how a model is saved (python api)
Now, we look at how the save/load works for saving actual model params, we look at the implementation of `save_vars` in fluid. [Code](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/io.py#L88). We see that a new program is created with `save` op is appended for each `vars` which is persistable. Then the executor runs this program.

#### Approach

We basically make two assumptions:
- For both load/save, the order of iterating over the variables is the same. (This should hopefully be true)
- We don't worry about the `overwrite` option which is in `save_op`.

While saving:
- We basically store a `uint64_t` number in addition to the actual serialized bytes as in the original `save`. This number will tell us about the size of the serialized LoDTensor in bytes.

- When the `save` is called for the first time, we will create a file, create a string that will have serialized LoDTensor data. Now we store the size of this string first in a fixed width (`uint64_t`) number, and then store the string.

- When the `save` is called later, we basically go to the end of the file, and store 2 things: the size of the string and the string itself.

While loading:
- We pass an additional attribute, in order to load the correct chunk of parameter. So we pass a counter value (which counts from 0 the relative order of the different params). 

- With this counter and the extra size information that we stored, we can hop to the appropriate part of the file, and read the chunk, and deserialize it.

For implementation, i think it will be better to have another op for this (rather than replacing the original save_op/load_op, so that is easier to debug, and i don't know the details of how the load_op and save_op are used in distributed version as of now).

**Comment by sidgoyal78** [comment]
I think we can even handle the "overwrite" file case. We can also have a counter for the `save` op, and if the counter is 0, and file exists and overwrite = false, we abort.
But if counter is 0, file exists, overwrite = true, we create a new file (open file in write mode).

**Comment by Xreki** [comment]
Let me take an example to explain the saving format as I understand.

If there are two parameters, `fc0.w0` and `fc0.b0`, they will be saved in a file as:

```text
uint64_t: size of fc0.w0
string: serialized LoDTensor data of fc0.w0
uint64_t: size of fc0.b0
string: serialized of LoDTensor data of fc0.b0
```

Am I right? I think we may need another string to record the parameter's name.

```text
string: fc0.w0
uint64_t: size of fc0.w0
string: serialized LoDTensor data of fc0.w0
string: fc0.b0
uint64_t: size of fc0.b0
string: serialized of LoDTensor data of fc0.b0
```

> For implementation, i think it will be better to have another op for this (rather than replacing the original save_op/load_op, so that is easier to debug, and i don't know the details of how the load_op and save_op are used in distributed version as of now).

In the first implementation, we may fill the parameter's Tensor in the `Load()` function directly.

**Comment by sidgoyal78** [comment]
Yes, you are right. I think for the 'name' of the param it is not currently stored. So It should be merged with the serialized LoDTensor data and then the size should be generated:

So I am thinking of it more like this:
```
uint64_t: size of fc0.w0 + size of string fc0.w0 + 1
string: lengthof(fc0.w0) + fc0.w0 + serialized LoDTensor data of fc0.w0
uint64_t: size of fc0.b0 + size of string fc0.b0 + 1
string: lengthof(fc0.b0) + fc0.b0 + serialized of LoDTensor data of fc0.b0
```
(Otherwise, we won't know the size of the string fc0.w0/fc0.b0 beforehand, so we merge it with the serialization of the LoDtensor, and then generate the size)

**Comment by Xreki** [comment]
OK. The design is based on the order. We need to make sure the loading order is totally the same as the saving order.

**Comment by Xreki** [comment]
Fixed by #7995 and #7909
